src/lib/util/internal/loggerHelper.js

```
@@ -0,0 +1,25 @@
+const log = require('pino')()
+
+function getClientInfoFromHeaders (request) {
+  if (request && request.headers && request.headers['data-sync-client-info']) {
+    const encoded = request.headers['data-sync-client-info']
+    let buf
+    try {
+      buf = Buffer.from(encoded, 'base64')
+    } catch (e) {
+      log.error('Unable decode base64 data-sync-client-info header provided by the client. Message: ' + e.message)
+      return undefined
+    }
+
+    const decoded = buf.toString('utf8')
+    try {
+      return JSON.parse(decoded)
+    } catch (e) {
```

@aliok I'm not 100% sure about this.

If a client sends a malformed header, I would consider that request to be invalid and I would send back a 400 `Bad Request` error back to the client. This implementation would swallow the error right?

---

I am not sure about that @darahayes.

Especially, when you think about the mutation case, the operation has already happened and doesn't matter if we return 400 to user because of bad headers.

We could move decoding operation in front of the resolver execution and that would make things work consistently. And, we would only do the decoding and rejection if audit log is enabled.

I am not sure how important it is to block the requests because the client info is not sent properly anyway.

@wei-lee @wtrocki give us your opinon please

---

@darahayes 

On the other hand, I am aware they're called **AUDIT** logs so they're security features to track things down and rejecting requests with no client info might make sense...

But still, clients might send random shit as we don't really validate.

So, I'm -1 on rejecting requests with no/improper client/device/app info

---

In most cases I think I would agree with @darahayes, but here things are a bit different, in terms of that the purpose of the function is for the audit log, and we should log every request (even if it may contain invalid data). So from this POV, I think I would agree with @aliok. We can return 400 code as well, but we do need to log this request for auditing.

---

I'm -1 on rejecting requests that don't contain client info. That would make for a terrible experience all round. I think no client info is fine.

I'm +100 on immediately rejecting requests that contain malformed data, whether that's incorrect base64 encoding or invalid JSON. Every request that comes in is logged by our request middleware, regardless of the presence of errors. This is a feature outside of our audit logging. However, I think it's a really bad idea to allow a request to be fulfilled if the data is malformed.

If you think about this in the context of my comment below (parsing the header as soon as the request comes in) then I think it makes sense that your bad request should never even make it to the resolver level, so a mutation/query wouldn't occur at all. We'll still get a log that the request came in and that it was unsuccessful.